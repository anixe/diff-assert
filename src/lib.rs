mod diff;
pub use self::diff::*;
use itertools::Itertools;
use std::str::Lines;

#[macro_export]
macro_rules! assert_dbg {
    ($expected: expr, $actual: expr) => {
        $crate::assert_diff!(
            format!("{:#?}", $expected),
            format!("{:#?}", $actual))
    };
    ($expected: expr, $actual: expr, $message: literal $(,$message_args: expr)*) => {
        $crate::assert_diff!(
            format!("{:#?}", $expected),
            format!("{:#?}", $actual),
            $message, $($message_args),*)
    }
}

#[macro_export]
macro_rules! assert_diff {
    ($expected: expr, $actual: expr) => {
        $crate::assert_diff!($expected, $actual, "Found differences")
    };
    ($expected: expr, $actual: expr, $message: literal $(,$message_args: expr)*) => {
        $crate::inner_assert_diff($expected.lines(), $actual.lines(), format!($message, $($message_args),*));
    };
}

#[doc(hidden)]
pub fn inner_assert_diff(expected: Lines, actual: Lines, msg_fmt: String) {
    let e: Vec<String> = expected.map(String::from).collect();
    let a: Vec<String> = actual.map(String::from).collect();
    let result = crate::diff_hunks(&e, &a, 3).unwrap();
    if !result.is_empty() {
        let mut msg = String::from("\n");
        msg += &msg_fmt;
        msg += "\n\n";

        msg += &result.into_iter().map(|s| s.to_string()).join("\n");

        panic!("{}", msg)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use itertools::Itertools;
    use test_case::test_case;

    #[test_case("A B C\nD E F", "A B D\nE F G")]
    #[test_case("A B C\nD E F", "A B D\nE F G\n1 2 3")]
    #[test_case(r#"|-------|-------------|-----------------------------------------|-----|---------------------------------|-----------|
| 24638 | Twin Room   | P1:3 A1:3 C0:2 FC0:9[0:1] MCA0:3[13:13] | DZ  | child_ages:["2:3","4:6","7:12"] |           |"#,
   r#"|-------|-------------|-----------------------------------------|-----|---------------------------|-----------|
| 24638 | Twin Room   | P1:3 A1:3 C0:2 FC0:9[0:3] MCA0:3[13:13] | DZ  | child_ages:["4:6","7:12"] |           |"#
    )]
    fn test_diff_hunks(left: &str, right: &str) {
        let left: Vec<String> = left.lines().map(String::from).collect();
        let right: Vec<String> = right.lines().map(String::from).collect();
        let hunks = diff_hunks(&left, &right, 3).expect("hunks");

        if !hunks.is_empty() {
            let hunks = hunks.into_iter().map(|s| format!("{}\n", s)).join("\n");
            insta::assert_display_snapshot!(hunks);
        }
    }

    #[test_case(
         "\n\u{1b}[1;4mLorem ipsum\u{1b}[0m\n\n\nExcepteur sint occaecat cupidatat non proident\n\n\u{1b}[7m1\u{1b}[0m\n",
         "\n\u{1b}[1;4mLorem ipsum\u{1b}[0m\n\n\nExcepteur sint occaecat cupidatat non proident\n\n\u{1b}[7m2\u{1b}[0m\n"
    )]
    #[test_case(
    "\nLorem ipsum\n\n\nExcepteur sint occaecat cupidatat non proident\n\n1\n",
    "\nLorem ipsum\n\n\nExcepteur sint occaecat cupidatat non proident\n\n2\n"
    )]
    #[test_case(
    "\nL\n\n\nE\n\n1\n",
    "\nL\n\n\nE\n\n2\n"
    )]
    #[test_case(
    "\n\n\n\n\n\n1",
    "\n\n\n\n\n\n2"
    )]
    fn overflow_test(a: &str, b: &str) {
        let left: Vec<String> = a.lines().map(String::from).collect();
        let right: Vec<String> = b.lines().map(String::from).collect();
        let hunks = diff_hunks(&left, &right, 3).expect("hunks");
        dbg!(&hunks);
    }

    #[test_case(
    "\nLorem \n\n\nipsum\n1\n2\n3\n4\n",
    "\nLorem \n\n\nipsun\n1\n2\n3\n4\n"
    )]
    #[test_case(
    "\nLorem \n\n\nipsum\n1\n2\n3\n",
    "\nLorem \n\n\nipsun\n1\n2\n3\n"
    )]
    #[test_case(
    "\nLorem \n\n\nipsum\n1\n",
    "\nLorem \n\n\nipsun\n1\n"
    )]
    #[test_case(
        concat!(
            "1\n2\n3\n4\n", // unchanged
            "foo\n", // changed
            "1\n2\n3\n4\n", // unchanged
            "bar\n", // changed
            "1\n2\n3\n4\n", // unchanged
        ),
        concat!(
            "1\n2\n3\n4\n", // unchanged
            "fou\n", // changed
            "1\n2\n3\n4\n", // unchanged
            "baz\n", // changed
            "1\n2\n3\n4\n", // unchanged
        )
    )]
    fn bad_diff_test(a: &str, b: &str) {
        let left: Vec<String> = a.lines().map(String::from).collect();
        let right: Vec<String> = b.lines().map(String::from).collect();
        let hunks = diff_hunks(&left, &right, 3).expect("hunks");
        insta::assert_debug_snapshot!(hunks);
    }

    #[test]
    #[should_panic]
    fn test() {
        let expected = "foo
        bar"
        .to_string();

        let actual = "foo
        foo"
        .to_string();

        assert_diff!(expected, actual);
    }

    #[test]
    #[should_panic]
    fn dbg_test() {
        let expected = ("Foo", "Bar");
        let actual = ("Foo", "foo");
        assert_dbg!(expected, actual);
    }
}
