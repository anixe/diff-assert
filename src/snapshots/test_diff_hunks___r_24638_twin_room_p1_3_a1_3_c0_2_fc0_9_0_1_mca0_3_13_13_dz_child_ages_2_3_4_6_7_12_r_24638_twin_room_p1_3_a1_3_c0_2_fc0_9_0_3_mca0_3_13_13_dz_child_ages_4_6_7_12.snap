---
source: src/lib.rs
expression: hunks
---
[2;44;30m... ...   @@ -1,2 +1,2 @@[0m
[40;31m001      [1m-[0m[40;31m[0m[40;31m[2m|[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m|[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m|[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m|[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m|[0m[40;31m[7m-[0m[40;31m[7m-[0m[40;31m[7m-[0m[40;31m[7m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[7m-[0m[40;31m[7m-[0m[40;31m[2m|[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m-[0m[40;31m[2m|[0m[40;31m[0m
[40;31m002      [1m-[0m[40;31m[0m[40;31m[2m|[0m[40;31m[2m [0m[40;31m[2m2[0m[40;31m[2m4[0m[40;31m[2m6[0m[40;31m[2m3[0m[40;31m[2m8[0m[40;31m[2m [0m[40;31m[2m|[0m[40;31m[2m [0m[40;31m[2mT[0m[40;31m[2mw[0m[40;31m[2mi[0m[40;31m[2mn[0m[40;31m[2m [0m[40;31m[2mR[0m[40;31m[2mo[0m[40;31m[2mo[0m[40;31m[2mm[0m[40;31m[2m [0m[40;31m[2m [0m[40;31m[2m [0m[40;31m[2m|[0m[40;31m[2m [0m[40;31m[2mP[0m[40;31m[2m1[0m[40;31m[2m:[0m[40;31m[2m3[0m[40;31m[2m [0m[40;31m[2mA[0m[40;31m[2m1[0m[40;31m[2m:[0m[40;31m[2m3[0m[40;31m[2m [0m[40;31m[2mC[0m[40;31m[2m0[0m[40;31m[2m:[0m[40;31m[2m2[0m[40;31m[2m [0m[40;31m[2mF[0m[40;31m[2mC[0m[40;31m[2m0[0m[40;31m[2m:[0m[40;31m[2m9[0m[40;31m[2m[[0m[40;31m[2m0[0m[40;31m[2m:[0m[40;31m[7m1[0m[40;31m[2m][0m[40;31m[2m [0m[40;31m[2mM[0m[40;31m[2mC[0m[40;31m[2mA[0m[40;31m[2m0[0m[40;31m[2m:[0m[40;31m[2m3[0m[40;31m[2m[[0m[40;31m[2m1[0m[40;31m[2m3[0m[40;31m[2m:[0m[40;31m[2m1[0m[40;31m[2m3[0m[40;31m[2m][0m[40;31m[2m [0m[40;31m[2m|[0m[40;31m[2m [0m[40;31m[2mD[0m[40;31m[2mZ[0m[40;31m[2m [0m[40;31m[2m [0m[40;31m[2m|[0m[40;31m[2m [0m[40;31m[2mc[0m[40;31m[2mh[0m[40;31m[2mi[0m[40;31m[2ml[0m[40;31m[2md[0m[40;31m[2m_[0m[40;31m[2ma[0m[40;31m[2mg[0m[40;31m[2me[0m[40;31m[2ms[0m[40;31m[2m:[0m[40;31m[2m[[0m[40;31m[2m"[0m[40;31m[7m2[0m[40;31m[7m:[0m[40;31m[7m3[0m[40;31m[7m"[0m[40;31m[7m,[0m[40;31m[7m"[0m[40;31m[2m4[0m[40;31m[2m:[0m[40;31m[2m6[0m[40;31m[2m"[0m[40;31m[2m,[0m[40;31m[2m"[0m[40;31m[2m7[0m[40;31m[2m:[0m[40;31m[2m1[0m[40;31m[2m2[0m[40;31m[2m"[0m[40;31m[2m][0m[40;31m[2m [0m[40;31m[2m|[0m[40;31m[2m [0m[40;31m[2m [0m[40;31m[2m [0m[40;31m[2m [0m[40;31m[2m [0m[40;31m[2m [0m[40;31m[2m [0m[40;31m[2m [0m[40;31m[2m [0m[40;31m[2m [0m[40;31m[2m [0m[40;31m[2m|[0m[40;31m[0m
[40;32m    001  [1m+[0m[40;32m[0m[40;32m[2m|[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m|[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m|[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m|[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m|[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m|[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m-[0m[40;32m[2m|[0m[40;32m[0m
[40;32m    002  [1m+[0m[40;32m[0m[40;32m[2m|[0m[40;32m[2m [0m[40;32m[2m2[0m[40;32m[2m4[0m[40;32m[2m6[0m[40;32m[2m3[0m[40;32m[2m8[0m[40;32m[2m [0m[40;32m[2m|[0m[40;32m[2m [0m[40;32m[2mT[0m[40;32m[2mw[0m[40;32m[2mi[0m[40;32m[2mn[0m[40;32m[2m [0m[40;32m[2mR[0m[40;32m[2mo[0m[40;32m[2mo[0m[40;32m[2mm[0m[40;32m[2m [0m[40;32m[2m [0m[40;32m[2m [0m[40;32m[2m|[0m[40;32m[2m [0m[40;32m[2mP[0m[40;32m[2m1[0m[40;32m[2m:[0m[40;32m[2m3[0m[40;32m[2m [0m[40;32m[2mA[0m[40;32m[2m1[0m[40;32m[2m:[0m[40;32m[2m3[0m[40;32m[2m [0m[40;32m[2mC[0m[40;32m[2m0[0m[40;32m[2m:[0m[40;32m[2m2[0m[40;32m[2m [0m[40;32m[2mF[0m[40;32m[2mC[0m[40;32m[2m0[0m[40;32m[2m:[0m[40;32m[2m9[0m[40;32m[2m[[0m[40;32m[2m0[0m[40;32m[2m:[0m[40;32m[7m3[0m[40;32m[2m][0m[40;32m[2m [0m[40;32m[2mM[0m[40;32m[2mC[0m[40;32m[2mA[0m[40;32m[2m0[0m[40;32m[2m:[0m[40;32m[2m3[0m[40;32m[2m[[0m[40;32m[2m1[0m[40;32m[2m3[0m[40;32m[2m:[0m[40;32m[2m1[0m[40;32m[2m3[0m[40;32m[2m][0m[40;32m[2m [0m[40;32m[2m|[0m[40;32m[2m [0m[40;32m[2mD[0m[40;32m[2mZ[0m[40;32m[2m [0m[40;32m[2m [0m[40;32m[2m|[0m[40;32m[2m [0m[40;32m[2mc[0m[40;32m[2mh[0m[40;32m[2mi[0m[40;32m[2ml[0m[40;32m[2md[0m[40;32m[2m_[0m[40;32m[2ma[0m[40;32m[2mg[0m[40;32m[2me[0m[40;32m[2ms[0m[40;32m[2m:[0m[40;32m[2m[[0m[40;32m[2m"[0m[40;32m[2m4[0m[40;32m[2m:[0m[40;32m[2m6[0m[40;32m[2m"[0m[40;32m[2m,[0m[40;32m[2m"[0m[40;32m[2m7[0m[40;32m[2m:[0m[40;32m[2m1[0m[40;32m[2m2[0m[40;32m[2m"[0m[40;32m[2m][0m[40;32m[2m [0m[40;32m[2m|[0m[40;32m[2m [0m[40;32m[2m [0m[40;32m[2m [0m[40;32m[2m [0m[40;32m[2m [0m[40;32m[2m [0m[40;32m[2m [0m[40;32m[2m [0m[40;32m[2m [0m[40;32m[2m [0m[40;32m[2m [0m[40;32m[2m|[0m[40;32m[0m


